File: packages/sdk-go/melaya/billing.go
// Billing API — subscription status, Stripe checkout/portal sessions, and
// public pricing plans.
//
// Maps to /api/v1/private/billing/* (auth) and /api/v1/billing/plans (public).
package melaya

import "context"

// BillingAPI wraps Stripe billing endpoints.
type BillingAPI struct {
	h *httpClient
}

// Subscription returns the caller's current Stripe subscription status and tier.
//
// GET /api/v1/private/billing/subscription
func (b *BillingAPI) Subscription(ctx context.Context) (*BillingSubscription, error) {
	data, err := b.h.get(ctx, "/api/v1/private/billing/subscription", nil)
	if err != nil {
		return nil, err
	}
	var v BillingSubscription
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// CreateCheckout creates a Stripe Checkout session for a tier upgrade.
// Returns a URL to redirect the user to.
//
// POST /api/v1/private/billing/checkout
func (b *BillingAPI) CreateCheckout(ctx context.Context, body map[string]interface{}) (*BillingCheckoutResult, error) {
	data, err := b.h.post(ctx, "/api/v1/private/billing/checkout", body)
	if err != nil {
		return nil, err
	}
	var v BillingCheckoutResult
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// CreatePortal creates a Stripe Customer Portal session for managing the subscription.
// Returns a URL to redirect the user to.
//
// POST /api/v1/private/billing/portal
func (b *BillingAPI) CreatePortal(ctx context.Context) (*BillingPortalResult, error) {
	data, err := b.h.post(ctx, "/api/v1/private/billing/portal", nil)
	if err != nil {
		return nil, err
	}
	var v BillingPortalResult
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// Plans returns public pricing plan details (price IDs for forge/bastion/citadel tiers).
//
// GET /api/v1/billing/plans (public)
func (b *BillingAPI) Plans(ctx context.Context) ([]BillingPlan, error) {
	data, err := b.h.get(ctx, "/api/v1/billing/plans", nil)
	if err != nil {
		return nil, err
	}
	var v []BillingPlan
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

File: packages/sdk/src/billing.ts
/**
 * Billing API — subscription status, Stripe checkout/portal sessions, and
 * public pricing plans.
 *
 * Maps to `/api/v1/private/billing/*` (authenticated) and `/api/v1/billing/plans`
 * (public). Read-only for most consumers; checkout and portal return redirect URLs.
 *
 * @example
 * ```ts
 * const sub = await melaya.billing.subscription();
 * console.log(sub.tier, sub.status);
 *
 * const { url } = await melaya.billing.createCheckout({ priceId: "price_..." });
 * // redirect user to url
 * ```
 */
import type { HttpClient } from "./client.js";
import type {
  BillingCheckoutResult,
  BillingPlan,
  BillingPortalResult,
  BillingSubscription,
} from "./platform-types.js";

export class BillingAPI {
  constructor(private readonly http: HttpClient) {}

  /** Get the caller's current Stripe subscription status and tier. */
  async subscription(): Promise<BillingSubscription> {
    return this.http.get<BillingSubscription>("/api/v1/private/billing/subscription");
  }

  /**
   * Create a Stripe Checkout session for a tier upgrade.
   * Returns a URL to redirect the user to.
   */
  async createCheckout(body: { priceId?: string; tier?: string }): Promise<BillingCheckoutResult> {
    return this.http.post<BillingCheckoutResult>("/api/v1/private/billing/checkout", body);
  }

  /**
   * Create a Stripe Customer Portal session for managing the subscription.
   * Returns a URL to redirect the user to.
   */
  async createPortal(): Promise<BillingPortalResult> {
    return this.http.post<BillingPortalResult>("/api/v1/private/billing/portal");
  }

  /** Return public pricing plan details (price IDs for forge/bastion/citadel tiers). */
  async plans(): Promise<BillingPlan[]> {
    return this.http.get<BillingPlan[]>("/api/v1/billing/plans");
  }
}
